pantheon/src/main/java/tech/pegasys/pantheon/Runner.java

```diff
@@ -98,20 +97,47 @@ public void close() throws Exception {
     networkRunner.stop();
     networkRunner.awaitStop();
 
-    exec.shutdown();
     try {
-      jsonRpc.ifPresent(service -> service.stop().join());
-      websocketRpc.ifPresent(service -> service.stop().join());
-      metrics.ifPresent(service -> service.stop().join());
+      jsonRpc.ifPresent(service -> waitForServiceToStop("jsonRpc", service.stop()));
+      websocketRpc.ifPresent(service -> waitForServiceToStop("websocketRpc", service.stop()));
+      metrics.ifPresent(service -> waitForServiceToStop("metrics", service.stop()));
     } finally {
       try {
-        exec.awaitTermination(Long.MAX_VALUE, TimeUnit.DAYS);
+        vertx.close();
       } finally {
-        try {
-          vertx.close();
-        } finally {
-          pantheonController.close();
-        }
+        pantheonController.close();
+      }
+    }
+  }
+
+  private void waitForServiceToStop(
+      final String serviceName, final CompletableFuture<?> stopFuture) {
+    try {
+      stopFuture.get(30, TimeUnit.SECONDS);
+    } catch (final InterruptedException e) {
+      LOG.debug("Interrupted while waiting for service to complete", e);
+      Thread.currentThread().interrupt();
+    } catch (final ExecutionException e) {
+      LOG.error("Service " + serviceName + " failed to shutdown", e);
+    } catch (final TimeoutException e) {
+      LOG.error("Service {} did not shut down cleanly", serviceName);
+    }
+  }
+
+  private void waitForServiceToStart(
+      final String serviceName, final CompletableFuture<?> startFuture) {
+    while (!startFuture.isDone()) {
+      try {
+        startFuture.get(60, TimeUnit.SECONDS);
+      } catch (final InterruptedException e) {
+        LOG.debug("Interrupted while waiting for service to start", e);
+        Thread.currentThread().interrupt();
+        return;
```

Comment:
Why return and not continue (retry)?  That would imply if the thread gets interrupted we don't care if it succeeds or fails to start up

Reply:
If we mark the thread as interrupted and retry it will immediately throw interrupted exception again when we call `get` and if we don't re-mark the thread as interrupted then we're just ignoring the interrupt.

I'm going to go with throwing an exception so that startup explicitly fails and for an obvious reason.  Must admit, I'm not entirely against just ignoring the interrupt in this case either...